include/aspect/particle/particle_handler.h

```diff
@@ -128,6 +139,50 @@ namespace aspect
         remove_particle(const particle_iterator &particle);
 
         /**
+         * Insert a particle into the collection of particles. Return an iterator
+         * to the new position of the particle.
+         */
+        particle_iterator
+        insert_particle(const Particle<dim,spacedim> &particle,
+                        const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell);
+
+        /**
+         * This function allows to register three additional functions that are
+         * called every time a particle is transferred to another process
+         * (i.e. during sorting into cells, during ghost particle transfer, or
+         * during serialization of all particles.
+         *
+         * @param size_callback A function that is called when serializing
+         * particle data. The function gets no arguments and is expected to
+         * return the size of the additional data that is serialized per
+         * particle. Note that this currently implies the data size has to be
+         * the same for every particle, but it does not have to be the same for
+         * every serialization process (e.g. a serialization during particle
+         * movement might include temporary data, while a serialization after
+         * movement was finished does not need to transfer this data).
```

Comment:
Out of curiosity, how is the callback supposed to know what is happening at the time it is called?